packages/fuchsia_remote_debug_protocol/lib/src/dart/dart_vm.dart

```diff
@@ -0,0 +1,180 @@
+// Copyright 2018 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:io';
+
+import 'package:json_rpc_2/json_rpc_2.dart' as json_rpc;
+import 'package:web_socket_channel/io.dart';
+
+import '../common/logging.dart';
+
+const Duration _kConnectTimeout = const Duration(seconds: 30);
+
+const Duration _kReconnectAttemptInterval = const Duration(seconds: 3);
+
+const Duration _kRpcTimeout = const Duration(seconds: 5);
+
+final Logger _log = new Logger('DartVm');
+
+/// Definition of an asynchronous function for connecting to a `Uri`.
+///
+/// Creates a JSON RPC-2 connection.
+typedef Future<json_rpc.Peer> RpcPeerConnectionFunction(Uri uri);
+
+/// `DartVm` uses this function to connect to the Dart VM on Fuchsia.
+///
+/// This function can be assigned to a different one out in the event that a
+/// custom connection function is needed.
+RpcPeerConnectionFunction fuchsiaVmServiceConnectionFunction = _waitAndConnect;
+
+/// Attempts to connect to a Dart VM service.
+///
+/// Gives up after `_kConnectTimeout` has elapsed.
+Future<json_rpc.Peer> _waitAndConnect(Uri uri) async {
+  final Stopwatch timer = new Stopwatch()..start();
+
+  Future<json_rpc.Peer> attemptConnection(Uri uri) async {
+    WebSocket socket;
+    json_rpc.Peer peer;
+    try {
+      socket = await WebSocket.connect(uri.toString());
+      peer = new json_rpc.Peer(new IOWebSocketChannel(socket).cast())..listen();
+      return peer;
+    } catch (e) {
+      await peer?.close();
+      await socket?.close();
+      if (timer.elapsed < _kConnectTimeout) {
+        _log.info('Attempting to reconnect');
+        await new Future<Null>.delayed(_kReconnectAttemptInterval);
+        return attemptConnection(uri);
+      } else {
+        _log.severe('Connection to Fuchsia\'s Dart VM timed out at '
+            '${uri.toString()}');
+        rethrow;
+      }
+    }
+  }
+
+  return attemptConnection(uri);
+}
+
+/// Restores the VM service connection function to the original state.
```

Comment:
nit: s/original state/default implementation/ (it's not clear what "state" means for a function).

Reply:
Done.